Feature/date picker

malcolmmadsheep opened the pull request:
@nirhart Can you, please, take a look at this?

lbelinsk commented on testkit/index.js:
```diff
@@ -167,3 +167,6 @@ export const popoverMenuTestkitFactory = testkitFactoryCreator(popoverMenuDriver
 import badgeDriverFactory from '../src/Badge/Badge.driver';
 export const badgeTestkitFactory = testkitFactoryCreator(badgeDriverFactory);
+
+import datePickerDriverFactory from '../src/DatePicker/DatePicker.driver';
```
Export the enzyme testkit from 'testkit/enzyme as well'

lbelinsk commented on src/DatePicker/DatePicker.driver.js:
```diff
@@ -1,51 +1,22 @@
+  const getElementByDataHook = dataHook => element.querySelector(`[data-hook="${dataHook}"]`);
+  const getDatePickerInput = () => element.querySelector('input');
+  const getDatePickerInputValue = () => getDatePickerInput().value;
+  const getDatePicker = () => element.querySelector('.react-datepicker');
```
getCalendar will be more accurate. Datepicker is the component's name including the input..

lbelinsk commented on src/DatePicker/DatePicker.spec.js:
```diff
@@ -1,57 +1,108 @@
+    });
+
+    it('should set datePicker disabled from props', () => {
+      const driver = createDriver(<DatePicker onChange={onChange} disabled={true}/>);
```
disabled is enough, no need to write disabled={true}

lbelinsk commented on src/DatePicker/DatePicker.spec.js:
```diff
@@ -1,57 +1,108 @@
+    it('should set datePicker disabled from props', () => {
+      const driver = createDriver(<DatePicker onChange={onChange} disabled={true}/>);
+
+      expect(driver.getDatePickerInput().hasAttribute('disabled')).toBeTruthy();
```
better to check that the calendar is not opened on click and that the input and use the inputTestkit.isDisabled() method for that..
https://github.com/wix/wix-style-react/blob/master/src/Input/Input.driver.js#L46

lbelinsk commented on src/DatePicker/DatePicker.spec.js:
```diff
@@ -1,57 +1,108 @@
+    it('should not render datePicker on start', () => {
+      const driver = createDriver(<DatePicker onChange={onChange}/>);
+
+      expect(driver.getDatePicker()).toBeFalsy();
```
Maybe we can add isVisible method.. might be very useful for other projects as well..

lbelinsk commented on src/DatePicker/DatePicker.spec.js:
```diff
@@ -1,57 +1,108 @@
+
+      expect(onChange).toHaveBeenCalled();
     });
 
```
What about selecting with keyboard press.. not only with mouse

lbelinsk commented on src/DatePicker/DatePicker.spec.js:
```diff
@@ -1,57 +1,108 @@
+      const option = driver.getDatePickerOption();
+      driver.clickOn(option);
+
+      expect(onChange).toHaveBeenCalled();
```
Can be more specific - toHaveBeenCalledWith(....)

lbelinsk commented on src/DatePicker/DatePicker.spec.js:
```diff
@@ -1,57 +1,108 @@
-      });
+      expect(driver.getDatePicker()).toBeFalsy();
+    });
+
```
This is a good test.. Add a similar one that checks that the calendar doesn't close if `shouldCloseOnSelect={false}` is provided..

malcolmmadsheep commented:
@nirhart Thanks for review. Yeap, it's a great idea to split functionality by several drivers and we will probably do it. We make this PR to ask about if we move the right way cause it's our first time with wix-style-react and first of all we wanted to adapt to this project and it's testing flow.

lbelinsk commented:
Yeah, I think it is good that you shared your progress, that is why I shared my thoughts as well..
Feel free to ask for more semi-reviews like this one whenever you feel like you need one..

nirhart commented:
@lbelinsk please merge this pr when you think its good enough

malcolmmadsheep commented:
@lbelinsk We've just pushed updated PR with split driver logic and updated tests according to it. Now we continue to work on other props, that we haven't covered yet. We also would like to know about what should be tested with e2e tests. We think that we should test here some key features like opening with focus on element or selecting date with keys. It seems like some tests may have the same logic as unit tests and we are confused about how should we split logic between the tests. WDYT about it?

lbelinsk commented on src/DatePicker/DatePicker.spec.js:
```diff
@@ -1,57 +1,234 @@
+      expect(inputDriver.getValue()).toBe('02/10/2017');
+    });
+
+    it('should set placeholder from props if value is not set', () => {
```
It is ok, but lets talk about it tomorrow, I am not sure if we want to support this.
I think maybe the best usage will be instead of doing:
```
<DatePicker placeHoderText="text"/> 
```

To do:
```
<DatePicker customInput={<Input placeholder="text"/>}/>
``` 

In this way we don't mix props fof input with props for datepicker
I am not sure, let's think together tomorrow about it..

lbelinsk commented on src/DatePicker/DatePicker.spec.js:
```diff
@@ -1,57 +1,234 @@
+      expect(inputDriver.hasPrefix()).toBe(true);
+    });
+
+    it('should show calendar on select with ArrowUp key', () => {
```
This test is great, but it shouldn't be under this describe IMHO..

lbelinsk commented on src/DatePicker/DatePicker.spec.js:
```diff
@@ -1,57 +1,234 @@
-      <DatePicker onChange={onChangeMock}/>,
-      {attachTo: datePickerRoot}
-    );
+      expect(calendarDriver.exists()).toBeTruthy();
```
You meant inputDriver, right?

lbelinsk commented on src/DatePicker/DatePicker.spec.js:
```diff
@@ -1,57 +1,234 @@
+
+      const newValue = onChange.mock.calls[0][0];
+
+      expect(onChange).toHaveBeenCalled();
```
I like better the syntax toHaveBeenCalledWith() instead of the additional line newValue.diff(expectedValue...

ihork replied:
But how do you suggest to compare objects? When we click on next day `calendarDriver.triggerOnNthDay({trigger: 'click', n: 1});` new moment object will be created. I can't see a problem with `onChange.mock.calls[0][0]` - it's actually the same. We even don't need `expect(onChange).toHaveBeenCalled();` in this case.

lbelinsk commented on src/DatePicker/DatePicker.spec.js:
```diff
@@ -1,57 +1,234 @@
+      expect(calendarDriver.isVisible()).toBe(false);
+    });
+
+    it('should close calendar on select with Enter key', () => {
```
What about Tab, Escape keys...

ihork replied:
Yeap, we wanted to start tests in all directions, and now we 're going to cover it with all use cases.

lbelinsk commented on src/DatePicker/DatePicker.spec.js:
```diff
@@ -1,57 +1,234 @@
+      expect(onChange).not.toHaveBeenCalled();
+      expect(calendarDriver.isVisible()).toBe(true);
     });
+  });
```
Tests are really great!
I have 2 more in mind:

1. Test that onSelect is not called if trying to select a selected value.. This is an optimization we try to keep in all components..

2. We didn't test the yearDropdown or the MonthDropdown at all..

ihork replied:
1. Correct, will be done
2. We thought about should we need to cover it in e2e or not

lbelinsk commented:
About the e2e tests, they should be in high level..
Test a basic usage, open calendar, select a date, change the date withe keyboard and mouse click.

No need to test if the calendar is opening on click, or other unit tests like this because it is already covered by the component tests you wrote..

Try to expose some useful methods in the protractor driver, even if you are not using them... but make sure they work though :)

ihork commented on src/DatePicker/DatePickerInput.js:
```diff
@@ -14,7 +12,8 @@ export default class DatePickerInput extends Component {
     onEnterPressed: PropTypes.func,
-    customInput: PropTypes.node
+    customInput: PropTypes.node,
+    dataHook: PropTypes.string
```
We are using `inputDataHook` property to pass it in customInput as dataHook property

ihork commented on src/DatePicker/DatePicker.js:
```diff
@@ -62,7 +65,7 @@ export default class DatePicker extends Component {
   render() {
     return (
-      <div className={css.wrapper}>
+      <div className={css.wrapper} data-hook={this.props.dataHook}>
```
Looks great, thanks

lbelinsk commented:
Failing on ci due to lint errors

lbelinsk commented:
@malcolmmadsheep I am merging this pr after the build will pass..
Don't forget to update the documentation for the updated testkit methods in the next pr..

lbelinsk commented on src/DatePicker/DatePicker.spec.js:
```diff
@@ -252,11 +252,32 @@ describe('DatePicker', () => {
+        const {calendarDriver, inputDriver} = createDriver(
+          <DatePicker
+            onChange={onChange}
+            showYearDropdown={true}
```
no need for the ={true}

lbelinsk commented:
You have again lint errors...
Please run yarn build before you push..

One more thing - dataHook={'datepicker-default'}`
Please change to 'story-datepicker-default' as this is the convention across the lib..

The build should be green after fixing the linting errors

ihork commented:
Yeap, sorry for that. We had started e2e and we weren't prepared to merge, so we pushed just to sync between us.

lbelinsk commented:
ok.. Let's try to make this branch green.. It is red from the beginning (mainly because of yoshi) so we will be able to merge after the e2e's are ready

lbelinsk commented on src/DatePicker/DatePicker.protractor.driver.js:
```diff
@@ -0,0 +1,33 @@
+
+  return {
+    input: {
+      exists: () => getInput().isPresent(),
```
I think that I would want a user to use it like: `datePickerDriverFactory({dataHook: data-hook}).inputDriver.getValue()`
So rename the input to inputDriver..

lbelinsk commented on src/DatePicker/DatePicker.protractor.driver.js:
```diff
@@ -0,0 +1,33 @@
+  return {
+    input: {
+      exists: () => getInput().isPresent(),
+      isDisplayed: () => getInput().isDisplayed(),
```
It is more common (at least in this lib) to call it isVisible instead of isDisplayed..

lbelinsk commented on src/DatePicker/DatePicker.e2e.js:
```diff
@@ -0,0 +1,137 @@
+
+    input.click();
+
+    expect(calendar.exists()).toBe(false);
```
isVisible instead of exists

lbelinsk commented on src/DatePicker/DatePicker.e2e.js:
```diff
@@ -0,0 +1,137 @@
+      expect(calendar.exists()).toBe(true);
+
+      input.pressEscKey();
+      expect(calendar.exists()).toBe(false);
```
isVisible

lbelinsk commented on src/DatePicker/DatePicker.e2e.js:
```diff
@@ -0,0 +1,137 @@
+      const {input, calendar} = testkit;
+
+      input.click();
+      expect(calendar.exists()).toBe(true);
```
isVisible

lbelinsk commented:
Build is red only because eyes.it fails on the new datePicker examples, so it is all good.. will merge after the master build will be green.
Thanks a lot  @ihork and @malcolmmadsheep really great work!

nirhart commented:
@lbelinsk
